maxitg/SetReplace
# New property names

## Changes
* Closes #85.
* Implemented new properties for `WolframModelEvolutionObject`:
  * `"FinalStatePlot"`: `WolframModelPlot`s the last state,
  * `"StatesPlotsList"`: yields a list of `WolframModelPlot`s,
  * `"PartialGenerationsCount"`: total - complete generations,
  * `"GenerationComplete"`: yields true/false depending whether the last (or specified) generation is complete,
  * `"GenerationEventsCountList"`: event counts in between generations,
  * `"GenerationEventsList"`: events grouped by generation,
  * `"VertexCountList"`: counts of vertices at each generation,
  * `"EdgeCountList"`: counts of edges at each generation,
  * `"EventsStatesList"`: riffled states and events.
* Changed behavior of `"GenerationsCount"`, which now returns a list `{"CompleteGenerationsCount", "PartialGenerationsCount"}`.
* Renamed properties (old names are still allowed):
  * `"UpdatedStatesList"` -> `"AllEventsStatesList"`,
  * `"AllExpressions"` -> `"AllEventsEdgesList"`,
  * `"CreatorEvents"` -> `"EdgeCreatorEventIndices"`,
  * `"DestroyerEvents"` -> `"EdgeDestroyerEventIndices"`,
  * `"MaxCompleteGeneration"` -> `"CompleteGenerationsCount"`,
  * `"EventGenerations"` -> `"AllEventsGenerationsList"`,
  * `"ExpressionGenerations"` -> `"EdgeGenerationsList"`,
  * `"EventsCount"` -> `"AllEventsCount"`,
  * `"EventsList"` -> `"AllEventsList"`,
  * `"AtomsCountFinal"` -> `"FinalDistinctElementsCount"`,
  * `"AtomsCountTotal"` -> `"AllEventsDistinctElementsCount"`,
  * `"ExpressionsCountFinal"` -> `"FinalEdgeCount"`,
  * `"ExpressionsCountTotal"` -> `"AllEventsEdgesCount"`,
  * `"SetAfterEvent"` -> `"StateAfterEvent"`

## Review questions
* The events in the output of `"EventsStatesList"` refer to indices of expressions in “AllExpressions” rather than to the states that surround them, which make it rather confusing to read and unusable without additional information. Do you think it would be better to reindex them so that they refer to the indices in surrounding states? Or is it better to keep events spec consistent for all properties? Another alternative would be to return states as well as lists of indices in “EventsStatesList”.
* Please look at the new property names, and comment if it's clear from the names what they do. Some of these names are quite confusing to me.
* @sskini2: This is the largest piece needed to make `WolframModel` consistent with documentation.

## Tests
* Plot final state directly without calling `WolframModelPlot`:
```
In[] := WolframModel[{{1, 1}} -> {{1, 2}, {2, 2}, {2, 2}}, {{1, 
   1}}, 6, "FinalStatePlot"]
```
![image](https://user-images.githubusercontent.com/1479325/73128669-d63f2080-3fa0-11ea-8dff-f97eaa495a38.png)
* Label vertices:
```
In[] := WolframModel[{{1, 1}} -> {{1, 2}, {2, 2}, {2, 2}}, {{1, 1}}, 
  3]["FinalStatePlot", VertexLabels -> Automatic]
```
![image](https://user-images.githubusercontent.com/1479325/73128674-e6570000-3fa0-11ea-8d94-4a38dd23f555.png)
* Plot all generations:
```
In[] := WolframModel[{{1, 1}} -> {{1, 2}, {2, 2}, {2, 2}}, {{1, 
   1}}, 5, "StatesPlotsList"]
```
![image](https://user-images.githubusercontent.com/1479325/73128681-fc64c080-3fa0-11ea-950b-1e3529fb02be.png)
* See how the number of incomplete generations increases with the events count:
```
In[] := ListPlot[WolframModel[{{1, 1}} -> {{1, 2}, {2, 2}, {2, 2}}, {{1, 
      1}}, <|"MaxEvents" -> #|>, "PartialGenerationsCount", 
    "EventOrderingFunction" -> "Random"] & /@ Range[200]]
```
![image](https://user-images.githubusercontent.com/1479325/73128697-42218900-3fa1-11ea-86c3-96aea802ff71.png)
* Check if a given generation is complete:
```
In[] := WolframModel[{{1, 1}} -> {{1, 2}, {2, 2}, {2, 2}}, {{1, 1}}, <|
  "MaxEvents" -> 1000|>, "EventOrderingFunction" -> "Random"]
```
![image](https://user-images.githubusercontent.com/1479325/73128719-99275e00-3fa1-11ea-87b8-afb0a14a9e3b.png)
```
In[] := %["GenerationComplete", 4]
```
```
Out[] = False
```
* Check how many events occur in between generations:
```
In[] := WolframModel[{{1, 1}} -> {{1, 2}, {2, 2}, {2, 2}}, {{1, 
   1}}, 10, "GenerationEventsCountList"]
```
```
Out[] = {1, 2, 4, 8, 16, 32, 64, 128, 256, 512}
```
* Show events grouped by generation:
```
In[] := WolframModel[{{1, 1}} -> {{1, 2}, {2, 2}, {2, 2}}, {{1, 1}}, 3, 
  "GenerationEventsList"] // Column
```
![image](https://user-images.githubusercontent.com/1479325/73128740-ddb2f980-3fa1-11ea-94ad-c1a96b2abf61.png)
* Check how the number of vertices grows with generations:
```
In[] := WolframModel[{{1, 1}} -> {{1, 2}, {2, 2}, {2, 2}}, {{1, 
   1}}, 10, "VertexCountList"]
```
```
Out[] = {1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024}
```
* Same for the number of edges:
```
In[] := WolframModel[{{1, 1}} -> {{1, 2}, {2, 2}, {2, 2}}, {{1, 
   1}}, 10, "EdgeCountList"]
```
```
Out[] = {1, 3, 7, 15, 31, 63, 127, 255, 511, 1023, 2047}
```
* Show events and states in the same list (edge indices are given in terms of `"AllEventsEdgesList"`):
```
In[] := WolframModel[{{1, 1}} -> {{1, 2}, {2, 2}, {2, 2}}, {{1, 1}}, 3, 
  "EventsStatesList"] // Grid
```
![image](https://user-images.githubusercontent.com/1479325/73204731-744ff980-410d-11ea-86e5-907555beacf5.png)
